From c8cc0f6e8919ac32762d4a30130b35e45d3efc26 Mon Sep 17 00:00:00 2001
From: Scott Mayhew <smayhew@redhat.com>
Date: Wed, 2 Jul 2014 19:22:52 -0400
Subject: [fs] nfs: Clear NFS_INO_REVAL_PAGECACHE when we update the file size

Message-id: <1404328973-44183-3-git-send-email-smayhew@redhat.com>
Patchwork-id: 85185
O-Subject: [RHEL 7.1 PATCH 2/3] [bz1114054] NFS: Clear NFS_INO_REVAL_PAGECACHE when we update the file size
Bugzilla: 1114054
RH-Acked-by: Steve Dickson <SteveD@redhat.com>
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>

From: Trond Myklebust <trond.myklebust@primarydata.com>

In nfs_update_inode(), if the change attribute is seen to change on
the server, then we set NFS_INO_REVAL_PAGECACHE in order to make
sure that we check the file size.
However, if we also update the file size in the same function, we
don't need to check it again. So make sure that we clear the
NFS_INO_REVAL_PAGECACHE that was set earlier.

Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
(cherry picked from commit f2467b6f64da7d8446211c21f6fcd8fa10e8d948)
Signed-off-by: Scott Mayhew <smayhew@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfs/inode.c b/fs/nfs/inode.c
index e1e7e34..2dd5dbd 100644
--- a/fs/nfs/inode.c
+++ b/fs/nfs/inode.c
@@ -1581,6 +1581,7 @@ static int nfs_update_inode(struct inode *inode, struct nfs_fattr *fattr)
    if ((nfsi->npages == 0) || new_isize > cur_isize) {
     i_size_write(inode, new_isize);
     invalid |= NFS_INO_INVALID_ATTR|NFS_INO_INVALID_DATA;
+    invalid &= ~NFS_INO_REVAL_PAGECACHE;
    }
    dprintk("NFS: isize change on server for file %s/%ld "
      "(%Ld to %Ld)\n",
-- 
1.7.1